FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1323 @@
+          Ref ResultNode {};
+          Ref Value = LoadStackValueAtTop_Slow();
+          // Negate value
+          if (ReducedPrecisionMode) {
+            ResultNode = IREmit->_VFNeg(8, 8, Value);
+          } else {
+            ResultNode = IREmit->_VXor(16, 1, Value, HelperNode);
+          }
+          StoreStackValueAtTop_Slow(ResultNode, false);
+        } else { // fast path
+          StackData.setTop(StackMemberInfo {
+            .SourceDataSize = StackMember.SourceDataSize,
+            .SourceDataNode = nullptr,
+            .StackDataNode = ReducedPrecisionMode ? IREmit->_VFNeg(8, 8, StackMember.StackDataNode) :
+                                                    IREmit->_VXor(16, 1, StackMember.StackDataNode, HelperNode),
+
+          });
+        }
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80STACKABS: {
+        const auto& [Valid, StackMember] = StackData.top();
+
+        Ref HelperNode {};
+        if (!ReducedPrecisionMode) {
+          // Intermediate insts
+          // FIXME: maybe we could just remove this OP and have a VAND Stack instead.
+          Ref Low = GetConstant(~0ULL);
+          Ref High = GetConstant(0b0'111'1111'1111'1111ULL);
+          HelperNode = IREmit->_VCastFromGPR(16, 8, Low);
+          HelperNode = IREmit->_VInsGPR(16, 8, 1, HelperNode, High);
+        }
+
+        MigrateToSlowPathIf(Valid != StackSlot::VALID);
+
+        if (SlowPath) {
+          // slow path
+          Ref Value = LoadStackValueAtTop_Slow();
+          Ref ResultNode {};
+          if (ReducedPrecisionMode) {
+            ResultNode = IREmit->_VFAbs(8, 8, Value);
+          } else {
+            ResultNode = IREmit->_VAnd(16, 1, Value, HelperNode);
+          }
+          StoreStackValueAtTop_Slow(ResultNode, false);
+        } else {
+          // fast path
+          StackData.setTop(StackMemberInfo {.SourceDataSize = StackMember.SourceDataSize,
+                                            .SourceDataNode = nullptr,
+                                            .StackDataNode = ReducedPrecisionMode ? IREmit->_VFAbs(8, 8, StackMember.StackDataNode) :
+                                                                                    IREmit->_VAnd(16, 1, StackMember.StackDataNode, HelperNode),
+                                            .InterpretAsFloat = StackMember.InterpretAsFloat});
+        }
+
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+
+      case OP_F80CMPSTACK: {
+        const auto* Op = IROp->C<IROp_F80CmpStack>();
+        auto offset = Op->SrcStack;
+        const auto& [Valid1, StackMember1] = StackData.top();
+        const auto& [Valid2, StackMember2] = StackData.top(offset);
+
+        MigrateToSlowPathIf(Valid1 != StackSlot::VALID || Valid2 != StackSlot::VALID);
+
+        Ref CmpNode {};
+        if (SlowPath) {
+          Ref StackValue1 = LoadStackValueAtTop_Slow();
+          Ref StackValue2 = LoadStackValueAtOffset_Slow(offset);
```

Comment:
Similarly the  LoadStackValue helper should take an optional (default = 0) offset so we can use it here too. Again, that either does the slow path or the fast top(offset). And of course that top(offset) call would migrate automatically. So putting it all together, this switch case will look like:


```cpp
      case OP_F80CMPSTACK: {
        const auto* Op = IROp->C<IROp_F80CmpStack>();
        const auto Value1 = LoadStackValue();
        const auto Value2 = LoadStackValue(Op->SrcStack);

        Ref CmpNode = ReducedPrecisionMode ? IREmit->_FCmp(8, Value1, Value2) : 
                                                                                   IREmit->_F80Cmp(Value1, Value2;
        IREmit->ReplaceUsesWithAfter(CodeNode, CmpNode, CodeNode);
        IREmit->Remove(CodeNode);
        break;
      }
```

...which I'm sure you'll agree is a lot nicer!